### PR TITLE
add torch.cosine_similarity

### PR DESCRIPTION
**Pull request by vieting**

### CODE REVIEW

**Comment by vieting** [comment]
So far, I added a `DotProductAlongDim` module to do the dot product. The rest is done inside `torch.nn.functional.cosine_similarity` using existing functions. However, there is a problem with the `DotLayer`:

```
    line: assert not set(b_reduce_axes).intersection(b_var_axes)
    locals:
      set = <builtin> <class 'set'>
      b_reduce_axes = <local> [2]
      intersection = <not found>
      b_var_axes = <local> [2]
```
for the layer
```
*** root/'DotProductAlongDim' layer dict: {'class': 'dot', 'red1': 'F', 'red2': 'F', 'from': ['data', 'add']}
```
with inputs
```
layer root/'data' output: Data{'data', [B,T|'time:data'[B],F|F'feature:data'(7)]}
layer root/'add' output: Data{'add_output', [B,T|'time:data'[B],F|F'feature:data'(7)]}
```

File: pytorch_to_returnn/torch/nn/functional.py [hunk 1]
```diff
@@ -785,6 +785,13 @@ def multi_head_attention_forward(
     return attn_output, None
 
 
+def cosine_similarity(x1: Tensor, x2: Tensor, dim: Optional[int] = 1, eps: Optional[float] = 1e-8) -> Tensor:
+  mod = modules.DotProductAlongDim(dim=dim)
+  numerator = mod(x1, x2)
+  denominator = (x1 ** 2).sum() * (x2 ** 2).sum()
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/functional.py:791]: This is missing `sqrt` on the denominator, right?

File: pytorch_to_returnn/torch/nn/modules/linear.py [hunk 2]
```diff
@@ -123,6 +122,41 @@ def create_returnn_layer_dict(self, *inputs: Tensor, **kwargs) -> Dict[str, Any]
     return {"class": "dot", "red1": red1, "red2": red2, "var1": var1, "var2": var2, "from": sources}
 
 
+class DotProductAlongDim(Module):
+  """
+  Maps to RETURNN DotLayer for a special case
+  """
+  is_original_torch_module = False
+
+  def __init__(self, dim: int):
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/linear.py:131]: I would rather call it `reduce_dim` or so.
But also, I wonder whether it is more useful to not have this argument in the `__init__` but directly in the `create_returnn_layer_dict`?

File: pytorch_to_returnn/torch/nn/modules/linear.py [hunk 3]
```diff
@@ -123,6 +122,41 @@ def create_returnn_layer_dict(self, *inputs: Tensor, **kwargs) -> Dict[str, Any]
     return {"class": "dot", "red1": red1, "red2": red2, "var1": var1, "var2": var2, "from": sources}
 
 
+class DotProductAlongDim(Module):
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/linear.py:125]: I think I would simply call it `Dot` or `GenericDot`, and allow for potential generalization later when needed.

File: pytorch_to_returnn/torch/nn/modules/linear.py [hunk 4]
```diff
@@ -123,6 +122,41 @@ def create_returnn_layer_dict(self, *inputs: Tensor, **kwargs) -> Dict[str, Any]
...
+    assert all(len(input_.shape) >= 2 for input_ in [input1, input2]), "not implemented otherwise"
+
+    shape = input1.shape
+    assert input2.shape == shape, "not implemented otherwise"  # could e.g. consider broadcasting axes
+    del shape[self.dim]
+    shape = tuple(shape)
+    returnn_axis_from_torch_axis = {i: i for i in range(len(shape))}
+    return shape, returnn_axis_from_torch_axis
+
+  def create_returnn_layer_dict(self, *inputs: Tensor, **kwargs) -> Dict[str, Any]:
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/linear.py:149]:
```suggestion
  def create_returnn_layer_dict(self, a: Tensor, b: Tensor, *, reduce_dim_a: int, reduce_dim_b: int) -> Dict[str, Any]:
```

File: pytorch_to_returnn/torch/nn/modules/linear.py [hunk 5]
```diff
@@ -123,6 +122,41 @@ def create_returnn_layer_dict(self, *inputs: Tensor, **kwargs) -> Dict[str, Any]
...
+  """
+  Maps to RETURNN DotLayer for a special case
+  """
+  is_original_torch_module = False
+
+  def __init__(self, dim: int):
+    super(DotProductAlongDim, self).__init__()
+    self.dim = dim
+
+  def _get_output_shape_from_returnn(self, inputs_flat: List[Tensor], layer: LayerBase
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/linear.py:135]: Do you need the custom `_get_output_shape_from_returnn` for this layer? I would have assumed that the base logic should work.

File: pytorch_to_returnn/torch/nn/modules/linear.py [hunk 6]
```diff
@@ -123,6 +122,41 @@ def create_returnn_layer_dict(self, *inputs: Tensor, **kwargs) -> Dict[str, Any]
...
+                                     ) -> Tuple[Tuple[int, ...], Dict[int, int]]:
+    assert len(inputs_flat) == 2
+    input1, input2 = inputs_flat
+    assert all(isinstance(input_, Tensor) for input_ in [input1, input2])
+    assert all(len(input_.shape) >= 2 for input_ in [input1, input2]), "not implemented otherwise"
+
+    shape = input1.shape
+    assert input2.shape == shape, "not implemented otherwise"  # could e.g. consider broadcasting axes
+    del shape[self.dim]
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/linear.py:144]: Did you test this code? This here would not work I think because `shape` is a tuple (thus not mutable). You need to convert it to `list` first.

File: pytorch_to_returnn/torch/nn/modules/linear.py [hunk 7]
```diff
@@ -122,38 +122,17 @@ def create_returnn_layer_dict(self, *inputs: Tensor, **kwargs) -> Dict[str, Any]
...
-
-  def create_returnn_layer_dict(self, *inputs: Tensor, **kwargs) -> Dict[str, Any]:
-    sources = [self._get_input_layer_name(source) for source in inputs]
-    assert len(sources) == 2
-
-    assert inputs[0].shape == inputs[1].shape
-    dim0 = self._get_input_axis_to_returnn(inputs[0], self.dim)
-    dim1 = self._get_input_axis_to_returnn(inputs[1], self.dim)
-    assert dim0 == dim1
+  def create_returnn_layer_dict(self, a: Tensor, b: Tensor, reduce_dim_a: int, reduce_dim_b: int) -> Dict[str, Any]:
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/linear.py:131]:
```suggestion
  def create_returnn_layer_dict(self, a: Tensor, b: Tensor, *, reduce_dim_a: int, reduce_dim_b: int) -> Dict[str, Any]:
```

File: pytorch_to_returnn/torch/nn/functional.py [hunk 8]
```diff
@@ -786,9 +786,8 @@ def multi_head_attention_forward(
 
 
 def cosine_similarity(x1: Tensor, x2: Tensor, dim: Optional[int] = 1, eps: Optional[float] = 1e-8) -> Tensor:
-  mod = modules.DotProductAlongDim(dim=dim)
-  numerator = mod(x1, x2)
-  denominator = (x1 ** 2).sum() * (x2 ** 2).sum()
+  numerator = modules.Dot()(x1, x2, dim, dim)
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/functional.py:789]:
```suggestion
  numerator = modules.Dot()(x1, x2, reduce_dim_a=dim, reduce_dim_b=dim)
```

File: pytorch_to_returnn/torch/nn/functional.py [hunk 9]
```diff
@@ -785,6 +785,12 @@ def multi_head_attention_forward(
     return attn_output, None
 
 
+def cosine_similarity(x1: Tensor, x2: Tensor, dim: Optional[int] = 1, eps: Optional[float] = 1e-8) -> Tensor:
+  numerator = modules.Dot()(x1, x2, dim, dim)
+  denominator = sqrt(pow(x1, 2).sum() * pow(x2, 2).sum())
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/functional.py:790]: Maybe using `rqdrt` instead and then multiplying this would be more stable?

File: pytorch_to_returnn/torch/nn/functional.py [hunk 10]
```diff
@@ -785,6 +785,12 @@ def multi_head_attention_forward(
     return attn_output, None
 
 
+def cosine_similarity(x1: Tensor, x2: Tensor, dim: Optional[int] = 1, eps: Optional[float] = 1e-8) -> Tensor:
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/functional.py:788]: You don't use the `eps`. This is probably an error. It probably is added to the input of `rsqrt`?

Also, is `eps=None` really allowed? Otherwise `Optional` is wrong.

**Comment by albertz** [comment]
Btw, you need to rebase this such that we get the new behavior version (#93).

File: pytorch_to_returnn/torch/nn/functional.py [hunk 11]
```diff
@@ -799,6 +803,12 @@ def multi_head_attention_forward(
...
+  numerator = modules.Dot()(x1, x2, reduce_dim_a=dim, reduce_dim_b=dim)
+  denominator = rsqrt(pow(x1, 2).sum() * pow(x2, 2).sum())
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/functional.py:808]: Maybe call it `inv_denominator` or so.

File: pytorch_to_returnn/torch/nn/functional.py (hunk 10 above)
Comment (vieting) [reply pytorch_to_returnn/torch/nn/functional.py:806]: It's not added, but the maximum of the norms and `eps` is taken, see https://pytorch.org/docs/stable/generated/torch.nn.CosineSimilarity.html

Comment (vieting) [reply pytorch_to_returnn/torch/nn/functional.py:806]: Do you think using `min(cat(...))` makes sense here? Then it has to be fixed for scalars I think.

Comment (albertz) [reply pytorch_to_returnn/torch/nn/functional.py]: Ok, anyway, we should do the same then.

File: pytorch_to_returnn/torch/nn/functional.py [hunk 12]
```diff
@@ -799,6 +807,12 @@ def multi_head_attention_forward(
     return attn_output, None
 
 
+def cosine_similarity(x1: Tensor, x2: Tensor, dim: Optional[int] = 1, eps: float=1e-8) -> Tensor:
+  numerator = modules.Dot()(x1, x2, reduce_dim_a=dim, reduce_dim_b=dim)
+  inv_denominator = min(cat([rsqrt(pow(x1, 2).sum() * pow(x2, 2).sum()), tensor(eps ** -1)]))
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/functional.py:812]: Use `minimum` instead.

File: pytorch_to_returnn/torch/nn/functional.py [hunk 13]
```diff
@@ -799,6 +807,12 @@ def multi_head_attention_forward(
     return attn_output, None
 
 
+def cosine_similarity(x1: Tensor, x2: Tensor, dim: Optional[int] = 1, eps: float=1e-8) -> Tensor:
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/functional.py:810]: Is `dim=None` really allowed?

File: pytorch_to_returnn/torch/nn/functional.py (hunk 12 above)
Comment (vieting) [reply pytorch_to_returnn/torch/nn/functional.py:812]: To which RETURNN layer would we map that?

Comment (albertz) [reply pytorch_to_returnn/torch/nn/functional.py:812]: Maybe simply via `EvalLayer` and `sources: [a, b], eval: "tf.minimum(source(0), source(1))"`

**Comment by vieting** [comment]
Apparently the output still does not match.

**Comment by albertz** [comment]
> Apparently the output still does not match.

This is strange. Can you check at what point it gets wrong? You could also calculate it by hand for some simple values.

File: pytorch_to_returnn/torch/nn/functional.py [hunk 14]
```diff
@@ -799,6 +810,12 @@ def multi_head_attention_forward(
     return attn_output, None
 
 
+def cosine_similarity(x1: Tensor, x2: Tensor, dim: int=1, eps: float=1e-8) -> Tensor:
+  numerator = modules.Dot()(x1, x2, reduce_dim_a=dim, reduce_dim_b=dim)
+  inv_denominator = minimum(rsqrt(pow(x1, 2).sum() * pow(x2, 2).sum()), tensor(eps ** -1))
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/functional.py:815]: The `sum` is wrong. You only want to reduce the `dim`.

**Review by albertz — APPROVED** [review]
